library/std/src/sys/uefi/mod.rs

```diff
@@ -0,0 +1,173 @@
+        (common::get_boot_services(), uefi::env::image_handle())
+    {
+        let _ = unsafe {
+            ((*boot_services.as_ptr()).exit)(
+                handle.as_ptr(),
+                r_efi::efi::Status::ABORTED,
+                0,
+                [0].as_mut_ptr(),
+            )
+        };
+    }
+
+    // In case SystemTable and SystemHandle cannot be reached, do things the Windows way
+    #[allow(unused)]
+    const FAST_FAIL_FATAL_APP_EXIT: usize = 7;
+    unsafe {
+        cfg_if::cfg_if! {
+            if #[cfg(any(target_arch = "x86", target_arch = "x86_64"))] {
+                core::arch::asm!("int $$0x29", in("ecx") FAST_FAIL_FATAL_APP_EXIT);
+            } else if #[cfg(all(target_arch = "arm", target_feature = "thumb-mode"))] {
+                core::arch::asm!(".inst 0xDEFB", in("r0") FAST_FAIL_FATAL_APP_EXIT);
+            } else if #[cfg(target_arch = "aarch64")] {
+                core::arch::asm!("brk 0xF003", in("x0") FAST_FAIL_FATAL_APP_EXIT);
+            } else {
+                core::intrinsics::abort();
+            }
+        }
+        core::intrinsics::unreachable();
+    }
+}
+
+// This function is needed by the panic runtime. The symbol is named in
+// pre-link args for the target specification, so keep that in sync.
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn __rust_abort() {
+    abort_internal();
+}
+
+#[inline]
+pub fn hashmap_random_keys() -> (u64, u64) {
+    unsafe { (get_random().unwrap_or(1), get_random().unwrap_or(2)) }
+}
+
+unsafe fn get_random() -> Option<u64> {
+    use r_efi::protocols::rng;
+
+    let mut buf = [0u8; 8];
+    let handles = common::locate_handles(rng::PROTOCOL_GUID).ok()?;
+    for handle in handles {
+        if let Ok(protocol) = common::open_protocol::<rng::Protocol>(handle, rng::PROTOCOL_GUID) {
+            let r = unsafe {
+                ((*protocol.as_ptr()).get_rng)(
+                    protocol.as_ptr(),
+                    crate::ptr::null_mut(),
+                    buf.len(),
+                    buf.as_mut_ptr(),
+                )
+            };
+            if r.is_error() {
+                continue;
+            } else {
+                return Some(u64::from_le_bytes(buf));
+            }
+        }
+    }
+    None
+}
+
+extern "C" {
+    fn main(argc: isize, argv: *const *const u8) -> isize;
+}
+
+// FIXME: Do not generate this in case of `no_main`
+#[no_mangle]
+pub unsafe extern "efiapi" fn efi_main(
+    handle: r_efi::efi::Handle,
+    st: *mut r_efi::efi::SystemTable,
+) -> r_efi::efi::Status {
+    if let (Some(system_table), Some(system_handle)) = (NonNull::new(st), NonNull::new(handle)) {
+        unsafe { uefi::env::init_globals(system_handle, system_table.cast()) };
+
+        match unsafe { main(0, crate::ptr::null()) } {
+            0 => r_efi::efi::Status::SUCCESS,
+            _ => r_efi::efi::Status::ABORTED, // Or some other status code
+        }
+    } else {
+        r_efi::efi::Status::ABORTED
```

Comment:
I would prefer asserting on those values to be non-NULL. Otherwise, this will be really hard to debug and it is clearly an ABI violation if this happens.